### PR TITLE
Implemented attention Unet

### LINKED ISSUE #443: Add Attention to 2D UNet
The attention module is only available for 3D UNet. Adding the option to have 2D attention UNet would make ivadomed more modular.

### PR DESCRIPTION
**Pull request by AshkanTaghipour**
The 2D attention module is added to the code.
the headlines of the changes are elaborated in the following:
1. The UnetGridGatingSignal2 is added due to dealing with 2D input.
2. the _GridAttentionBlockND is rewrite based on the original code posted on GitHub (https://github.com/ozan-oktay/Attention-Gated-Networks) and now is compatible with 2D and 3D input.
3. In the Decoder class, the attention module is added.
4. some minor changes in the decoder and encoder class have been made due to make the code more modular (e.g., with depth changes, etc.)

Fixes #443

### CODE REVIEW

File: ivadomed/models.py [hunk 1]
```diff
@@ -383,14 +380,19 @@ class Decoder(Module):
         softmax (Softmax): Softmax layer that can be applied as last layer.
     """
 
+
     def __init__(self, out_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1,
-                 n_metadata=None, film_layers=None, hemis=False, final_activation="sigmoid", is_2d=True, n_filters=64):
+                 n_metadata=None, film_layers=None, hemis=False, final_activation="sigmoid", is_2d=True,
+                 n_filters=64, attention= True):
```
Comment (jcohenadad) [inline ivadomed/models.py:386]: please follow PEP8
```suggestion
                 n_filters=64, attention=True):
```

File: ivadomed/models.py [hunk 2]
```diff
@@ -325,14 +324,13 @@ def __init__(self, in_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1, n_meta
         self.down_path.append(max_pool(2))
 
         for i in range(depth - 1):
-            self.down_path.append(DownConv(n_filters, n_filters * 2, drop_rate, bn_momentum, is_2d))
-            self.down_path.append(FiLMlayer(n_metadata, n_filters * 2) if film_layers and film_layers[i + 1] else None)
+            self.down_path.append(DownConv(n_filters*2**i, n_filters*2**(i+1), drop_rate, bn_momentum, is_2d))
```
Comment (jcohenadad) [inline ivadomed/models.py:327]: This changes the behaviour. What is the justification? Maybe we should add a unit test to cover these lines of code?

Comment (AshkanTaghipour) [reply ivadomed/models.py:327]: @jcohenadad 
I think that it doesn't change the behavior, just the counter (n_filters) replaced by the counter of the loop for fewer lines of code.

File: ivadomed/models.py [hunk 3]
```diff
@@ -325,14 +324,13 @@ def __init__(self, in_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1, n_meta
...
+            self.down_path.append(FiLMlayer(n_metadata, n_filters * 2**(i+1)) if film_layers and film_layers[i + 1] else None)
             self.down_path.append(max_pool(2))
-            n_filters = n_filters * 2
 
         # Bottom
-        self.conv_bottom = DownConv(n_filters, n_filters, drop_rate, bn_momentum, is_2d)
```
Comment (AshkanTaghipour) [inline ivadomed/models.py:334]: In the U-net paper, in the bottom layer, the number of filters doubled but I think in this line the number of filters (output channels) remained the same as the input one.

File: ivadomed/models.py (hunk 2 above)
Comment (jcohenadad) [reply ivadomed/models.py:327]: but what about the `film_layers` condition?

Comment (AshkanTaghipour) [reply ivadomed/models.py:327]: about testing the validity of the code, I checked the output of each layer by entering an input tensor to the code and the related Jupyter notebook file can be found in the following link:
https://github.com/AshkanTaghipour/Attention_2D_Unet/blob/main/2D_attention_Unet.ipynb

however, any specific link or related subject about how to add a unit test would be appreciated.

Comment (jcohenadad) [reply ivadomed/models.py:327]: here: https://github.com/ivadomed/ivadomed/wiki/tests

Comment (AshkanTaghipour) [reply ivadomed/models.py:327]: about film_layers condition, the logic is the same, just similar to the previous version, the second argument of the FiLMlayer should have the size of the second argument of DownConv which it has.

Comment (AshkanTaghipour) [reply ivadomed/models.py:327]: The Unit test will be added.

File: ivadomed/models.py [hunk 4]
```diff
@@ -373,6 +369,7 @@ class Decoder(Module):
         film_layers (list): List of 0 or 1 indicating on which layer FiLM is applied.
         hemis (bool): Boolean indicating if HeMIS is on or not.
         final_activation (str): Choice of final activation between "sigmoid", "relu" and "softmax"
+        attention (bool): Boolian indicating if attention module is on or not.
```
Comment (charleygros) [inline ivadomed/models.py:372]:
```suggestion
        attention (bool): Boolean indicating if attention module is on or not.
```

File: ivadomed/models.py (hunk 1 above)
Comment (charleygros) [reply ivadomed/models.py:386]:
```suggestion
                 n_filters=64, attention=False):
```

File: ivadomed/models.py [hunk 5]
```diff
@@ -425,14 +428,36 @@ def __init__(self, out_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1,
...
+            else:
+                x = self.up_path[2*i](x, features[(self.depth-1)-i])
+            if self.up_path[2*i+1]:
+                x, w_film = self.up_path[2*i+ 1](x, context, w_film)
+                
 
-        for i in reversed(range(self.depth)):
-            x = self.up_path[-(i + 1) * 2](x, features[i])
-            if self.up_path[-(i + 1) * 2 + 1]:
-                x, w_film = self.up_path[-(i + 1) * 2 + 1](x, context, w_film)
```
Comment (charleygros) [inline ivadomed/models.py:435]: Looks similar to what was implemented before. You are just playing with different indexes. It's okay, but, when not necessary, I would recommend to keep the current way of indexes.

File: ivadomed/models.py [hunk 6]
```diff
@@ -476,6 +501,8 @@ class Unet(Module):
         final_activation (str): Choice of final activation between "sigmoid", "relu" and "softmax".
         is_2d (bool): Indicates dimensionality of model: True for 2D convolutions, False for 3D convolutions.
         n_filters (int):  Number of base filters in the U-Net.
+        attention (bool): Boolian indicating if attention module is on or not.
```
Comment (charleygros) [inline ivadomed/models.py:504]:
```suggestion
        attention (bool): Boolean indicating if attention module is on or not.
```

File: ivadomed/models.py [hunk 7]
```diff
@@ -488,12 +515,14 @@ def __init__(self, in_channel=1, out_channel=1, depth=3, drop_rate=0.4, bn_momen
         super(Unet, self).__init__()
 
         # Encoder path
-        self.encoder = Encoder(in_channel=in_channel, depth=depth, drop_rate=drop_rate, bn_momentum=bn_momentum,
-                               is_2d=is_2d, n_filters=n_filters)
+        self.encoder = Encoder(in_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1, n_metadata=None, film_layers=None,
+                 is_2d=True, n_filters=64)
```
Comment (charleygros) [inline ivadomed/models.py:519]: Why are you hard-coding here? eg we may want to have a multichannel model --> `in_channel` > 1

File: ivadomed/models.py [hunk 8]
```diff
@@ -488,12 +515,14 @@ def __init__(self, in_channel=1, out_channel=1, depth=3, drop_rate=0.4, bn_momen
...
 
         # Decoder path
-        self.decoder = Decoder(out_channel=out_channel, depth=depth, drop_rate=drop_rate, bn_momentum=bn_momentum,
-                               final_activation=final_activation, is_2d=is_2d, n_filters=n_filters)
+        self.decoder = Decoder(out_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1,
+                 n_metadata=None, film_layers=None, hemis=False, final_activation="sigmoid", is_2d=True,
+                 n_filters=64, attention=True)
```
Comment (charleygros) [inline ivadomed/models.py:524]: Same comment re hard coding.

File: ivadomed/models.py [hunk 9]
```diff
@@ -1159,25 +1190,22 @@ class GridAttentionBlockND(nn.Module):
 
     """
     def __init__(self, in_channels, gating_channels, inter_channels=None, dimension=3,
-                 sub_sample_factor=(2, 2, 2)):
-        super(GridAttentionBlockND, self).__init__()
+                 sub_sample_factor=2):
```
Comment (charleygros) [inline ivadomed/models.py:1193]: Could you please explain / comment this change?

File: ivadomed/models.py [hunk 10]
```diff
@@ -1205,21 +1233,25 @@ def __init__(self, in_channels, gating_channels, inter_channels=None, dimension=
...
+            weights_init_kaiming(m)
+         
         # Define the operation
         self.operation_function = self._concatenation
 
     def forward(self, x, g):
+        '''
+        :param x: (b, c, t, h, w)
+        :param g: (b, g_d)
+        :return:
```
Comment (charleygros) [inline ivadomed/models.py:1252]: Thanks for adding this docstring. Would be great if you could further detail, eg c = channel + match the docstring style used in `ivadomed` repository.

File: ivadomed/models.py [hunk 11]
```diff
@@ -425,14 +428,36 @@ def __init__(self, out_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1,
         self.last_conv = conv(in_channel // 2, out_channel, kernel_size=3, padding=1)
         self.last_film = FiLMlayer(n_metadata, self.out_channel) if film_layers and film_layers[-1] else None
         self.softmax = nn.Softmax(dim=1)
+        
+        ### ATTENTION MODULE ###
+        if self.attention:
+            
+            self.gating = UnetGridGatingSignal2(self.base_n_filter * 2**(self.depth),
```
Comment (andreanne-lemay) [inline ivadomed/models.py:435]: Decoder has an attribute `is_2d` if this attribute is set to false, we expect to use a 3D version of the UNet. Here,  `UnetGridGatingSignal2` should be `UnetGridGatingSignal3` if `is_2d` is set to False.

**Review by andreanne-lemay — COMMENTED** [review]
The code looks great in general (mostly PEP8 related comments) and I was able to run the attention unet for 2D UNet and Modified3DUNet :100: . One use case (regular 3D UNet) is not covered (when is_2d is set to false), see comment below. Overall excellent work! Thank you for this contribution!

File: model_test_attention2d [hunk 12]
```diff
@@ -0,0 +1,86 @@
+import ivadomed.models as imed_model
```
Comment (jcohenadad) [inline model_test_attention2d:1]: why duplicating the file and putting it under the root of the package? this is where the file should go: https://github.com/ivadomed/ivadomed/blob/master/testing/test_model.py

File: ivadomed/models.py [hunk 13]
```diff
@@ -293,7 +293,6 @@ def forward(self, x, y):
         x = self.downconv(x)
         return x
 
-
```
[the remark this answers is not in the document]
Comment (AshkanTaghipour) [reply ivadomed/models.py:296]: @andreanne-lemay many thanks for your valuable suggestions, about the PEP8 related issues, I am using Jupyter notebook but will migrate to pyCharm for future works ✌️ .

File: model_test_attention2d (hunk 12 above)
Comment (AshkanTaghipour) [reply model_test_attention2d]: sorry for that, I changed it into the test directory of the patch-1 branch, where the update_model is located.

Comment (jcohenadad) [reply model_test_attention2d]: c3b400e doesn't solve the problem: the file is now duplicated under the testing/ folder (!)
- https://github.com/ivadomed/ivadomed/pull/552/files?file-filters%5B%5D=No+extension#diff-74343adf7b5702ca368379078d5115777493c2e5dab9ab313421d6e191e052b9
- https://github.com/ivadomed/ivadomed/pull/552/files?file-filters%5B%5D=No+extension#diff-f32ddf02dd4be1853865f9a7662cd1681e02f844066be5e3e4740b9d79b32589

Please read how "pytest" works and make the appropriate changes

File: testing/test_model.py [hunk 14]
```diff
@@ -52,5 +52,35 @@ def test_film_generator():
...
+         torch.zeros((1, 256, 80, 80)), torch.zeros((1, 512, 40, 40))] # by changing the depth, the length of the input feature list should be changed.
+    inp = a # should not be transformed to torch tensor because "a" is the appended feature list from the skip connections.
+    model = imed_model.Decoder(out_channel=1, depth=3, drop_rate=0.4, bn_momentum=0.1,
+                 n_metadata=None, film_layers=None, hemis=False, final_activation="sigmoid", is_2d=True,
+                 n_filters=64, attention=True)
+    inf = model(inp)
+    assert(type(inf) == torch.Tensor)
+    
+def test_unet_att2d():
+
```
Comment (jcohenadad) [inline testing/test_model.py:76]:
```suggestion
```

File: testing/test_model.py [hunk 15]
```diff
@@ -52,5 +52,35 @@ def test_film_generator():
     assert (type(inf[0]) == torch.Tensor)
     assert (type(inf[1]) == torch.nn.parameter.Parameter)
 
+def test_modified_encoder():
+    # verifying changes in the Encoder class
```
Comment (jcohenadad) [inline testing/test_model.py:56]: is it really what this test verifies? currently it only checks if `inf` is a `tuple`

Comment (AshkanTaghipour) [reply testing/test_model.py:56]: yes, it was strange for me at first, but if we attend to the output of the encoder, we will find out that we need a list that contains features that represent skip connection of Unet network. 
furthermore, at the final line of code in the encoder class, a film layer joins this list, and these two features compacted as a tuple.

Comment (jcohenadad) [reply testing/test_model.py:58]: > if we attend to the output of the encoder, we will find out that we need a list that contains features that represent skip connection of Unet network.
furthermore, at the final line of code in the encoder class, a film layer joins this list, and these two features compacted as a tuple.

right, but at the end you are only assessing if your output is a tuple, your are not assessing its _integrity_

Comment (jcohenadad) [reply testing/test_model.py:58]: note: the same comment applies to your 2 other tests

Comment (AshkanTaghipour) [reply testing/test_model.py:58]: I think that these three tests evaluate the  _integrity_  assessment too. I added the unit test for encoder and decoder sections separately to ensure that some changes made by me will not change the behavior of the codebase. In the third unit test which is *unet_att_2d* the _integrity_ assessment is done because it consists of the encoder, decoder, and attention module.

using the separate unit test for encoder and decoder class helped me to find out in the bottom layer of the encoder section, the number of filters did not double in the [bottom layer](https://github.com/ivadomed/ivadomed/pull/552#discussion_r536899063)

File: ivadomed/models.py [hunk 16]
```diff
@@ -1195,7 +1223,7 @@ def __init__(self, in_channels, gating_channels, inter_channels=None, dimension=
             bn = nn.BatchNorm2d
             self.upsample_mode = 'bilinear'
         else:
-            raise NotImplementedError
+            raise NotImplemented
```
Comment (jcohenadad) [inline ivadomed/models.py:1226]: what is the reason for changing that?

Comment (AshkanTaghipour) [reply ivadomed/models.py:1226]: these lines come from the source code of [GridAttentionBlock] (https://github.com/ozan-oktay/Attention-Gated-Networks/blob/eee4881fdc31920efd873773e0b744df8dacbfb6/models/layers/grid_attention_layer.py#L44) which I think they want to expand their attention gate block, or for assessing their code. 
i think that the previous person who implemented the 3D attention module left this line intentionally as a reminder to implement the 2D or maybe check regularly the attention Gate source code for its additional feature.

File: testing/test_model.py (hunk 15 above)
Comment (AshkanTaghipour) [reply testing/test_model.py:58]: @jcohenadad my current problem is that I don't understand the error messages which surprisingly were crystal clear last night 😃, 
here are current error messages:
*FAILED testing/test_HeMIS.py::test_HeMIS - RuntimeError: Given groups=1, weig...
*FAILED testing/test_training_time.py::test_unet_time[config1-target_lst0-train_lst0]
*FAILED testing/test_training_time.py::test_unet_time[config5-target_lst0-train_lst0]
So, now I don't have any idea about why these errors show up while I just modify the test_model.py not test_HeMIS nor the test_training_time

File: ivadomed/models.py (hunk 16 above)
Comment (jcohenadad) [reply ivadomed/models.py:1226]: https://help.semmle.com/wiki/display/PYTHON/NotImplemented+is+not+an+Exception

File: testing/test_model.py (hunk 15 above)
Comment (jcohenadad) [reply testing/test_model.py:58]: I suggest you run pytest locally under the debugger

Comment (AshkanTaghipour) [reply testing/test_model.py:58]: I tested it locally and upload the related Jupiter notebook file [here](https://github.com/AshkanTaghipour/Attention_2D_Unet/blob/main/2D_attention_Unet.ipynb) that shows the validity of the 2d attention module on Unet. 
However, I will do it on PyCharm too.

File: ivadomed/models.py (hunk 16 above)
Comment (AshkanTaghipour) [reply ivadomed/models.py:1226]: > https://help.semmle.com/wiki/display/PYTHON/NotImplemented+is+not+an+Exception

👍

File: testing/test_HeMIS.py [hunk 17]
```diff
@@ -54,7 +54,7 @@ def test_HeMIS(p=0.0001):
             "name": "HeMISUnet",
             "dropout_rate": 0.3,
             "bn_momentum": 0.9,
-            "depth": 2,
+            "depth": 3,
```
Comment (jcohenadad) [inline testing/test_HeMIS.py:57]: why changing this behaviour? how is that related to the purpose of the PR?

Comment (AshkanTaghipour) [reply testing/test_HeMIS.py:57]: I don't have any errors in test_model.py and errors are in test_HeMIS.py and test_training_time, 
the first error message from  test_HeMIS.py is " weight of size [256, 512, 1, 1], expected input[4, 1024, 6, 6] to have *512* channels, but got *1024* channels instead"

I think that the main problem comes from [here](https://github.com/ivadomed/ivadomed/pull/552#discussion_r536899063) which the number of filters should be doubled but didn't, So, for the HeMIS condition, the number of input channel is twice the [last layer in the encoder] section(https://github.com/AshkanTaghipour/ivadomed/blob/d24e1a8d6fab17b8f62cee8adc5cd34ffbb2f1dd/ivadomed/models.py#L401) --> 2*512=1024
and the depth layer for HeMIS condition considered 2 which is [3 fact](https://github.com/AshkanTaghipour/ivadomed/blob/45011096a54750f1ba073bb10ed7d0bd60f08fb6/testing/test_HeMIS.py#L100)

Comment (jcohenadad) [reply testing/test_HeMIS.py:57]: well, I wouldn't change the condition of the test so the test can pass with your changes (ie: your changes should not impact the behaviour of other tests). Or did i miss something @charleygros @andreanne-lemay ?

Comment (AshkanTaghipour) [reply testing/test_HeMIS.py:57]: and I have another question, I read the testing folder of the Ivadomed, as an example, for the 'test_training_time' the parameters of the models have been set such as [here](https://github.com/AshkanTaghipour/ivadomed/blob/45011096a54750f1ba073bb10ed7d0bd60f08fb6/testing/test_training_time.py#L86), so when we want to do some changes in test_model.py as I did, and set the model parameters different from [here](https://github.com/AshkanTaghipour/ivadomed/blob/45011096a54750f1ba073bb10ed7d0bd60f08fb6/testing/test_training_time.py#L86) may cause some errors in other tests?
@charleygros @andreanne-lemay

**Comment by jcohenadad** [comment]
@AshkanTaghipour what it the status of this PR and how does the purpose differ from #545 ? Please avoid any duplication, thanks

**Comment by AshkanTaghipour** [comment]
> @AshkanTaghipour what it the status of this PR and how does the purpose differ from #545 ? Please avoid any duplication, thanks

the [#545](https://github.com/ivadomed/ivadomed/pull/545) should be closed. all modifications in `models.py` and adding test units in `test_model.py` have been made in [#552](https://github.com/ivadomed/ivadomed/pull/552)

**Comment by jcohenadad** [comment]
sorry i closed the wrong one 😬
@AshkanTaghipour if a PR should be closed, you have the responsibility to close it and justify why you close it (i do not have the time to do maintenance...)

**Comment by AshkanTaghipour** [comment]
> sorry i closed the wrong one 😬
> @AshkanTaghipour if a PR should be closed, you have the responsibility to close it and justify why you close it (i do not have the time to do maintenance...)

Ok. I will do it from now on.
